From cf71625b45bf51ea17f812a936f439a6d4ffbc57 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Thu, 19 Jun 2025 17:17:34 +0800
Subject: [PATCH 138/293] user32: Add ScheduleDispatchNotification() stub.

React Native apps use user32.2582 and expect the return value to be non-zero.
---
 dlls/user32/tests/input.c   |  27 +++
 dlls/user32/user32.spec     |   2 +-
 dlls/win32u/input.c         |  13 ++
 dlls/win32u/main.c          |   5 +
 dlls/win32u/win32syscalls.h | 322 ++++++++++++++++++------------------
 dlls/win32u/win32u.spec     |   2 +-
 dlls/wow64win/user.c        |   7 +
 include/ntuser.h            |   1 +
 8 files changed, 217 insertions(+), 162 deletions(-)

diff --git a/dlls/user32/tests/input.c b/dlls/user32/tests/input.c
index 4d4c3c81000..a82ca5941a2 100644
--- a/dlls/user32/tests/input.c
+++ b/dlls/user32/tests/input.c
@@ -430,6 +430,7 @@ static UINT (WINAPI *pGetRawInputDeviceInfoW) (HANDLE, UINT, void *, UINT *);
 static UINT (WINAPI *pGetRawInputDeviceInfoA) (HANDLE, UINT, void *, UINT *);
 static BOOL (WINAPI *pIsWow64Process)(HANDLE, PBOOL);
 static HKL (WINAPI *pLoadKeyboardLayoutEx)(HKL, const WCHAR *, UINT);
+static INT (WINAPI *pScheduleDispatchNotification)(HWND);
 
 /**********************adapted from input.c **********************************/
 
@@ -439,6 +440,8 @@ static void init_function_pointers(void)
 {
     HMODULE hdll = GetModuleHandleA("user32");
 
+    pScheduleDispatchNotification = (void *)GetProcAddress(hdll, (LPCSTR)2582);
+
 #define GET_PROC(func) \
     if (!(p ## func = (void*)GetProcAddress(hdll, #func))) \
       trace("GetProcAddress(%s) failed\n", #func)
@@ -6308,6 +6311,29 @@ static void test_SetFocus_process(void)
     DestroyWindow( hwnd );
 }
 
+static void test_ScheduleDispatchNotification(void)
+{
+    HWND hwnd;
+    INT ret;
+
+    if (!pScheduleDispatchNotification)
+    {
+        win_skip("ScheduleDispatchNotification is unavailable.\n");
+        return;
+    }
+
+    hwnd = CreateWindowW(L"static", NULL, WS_POPUP | WS_VISIBLE, 100, 100, 200, 200, NULL, NULL,
+                         NULL, NULL);
+
+    ret = pScheduleDispatchNotification(NULL);
+    ok(!ret, "Got unexpected %d.\n", ret);
+
+    ret = pScheduleDispatchNotification(hwnd);
+    ok(ret == 2, "Got unexpected %d.\n", ret);
+
+    DestroyWindow(hwnd);
+}
+
 START_TEST(input)
 {
     char **argv;
@@ -6352,6 +6378,7 @@ START_TEST(input)
     test_OemKeyScan();
     test_rawinput(argv[0]);
     test_DefRawInputProc();
+    test_ScheduleDispatchNotification();
 
     if(pGetMouseMovePointsEx)
         test_GetMouseMovePointsEx( argv );
diff --git a/dlls/user32/user32.spec b/dlls/user32/user32.spec
index b6e85a37982..732ab323414 100644
--- a/dlls/user32/user32.spec
+++ b/dlls/user32/user32.spec
@@ -93,7 +93,7 @@
 2579 stub -noname SetWindowShowState  # NtUserSetWindowShowState
 
 2581 stub -noname GetWindowTrackInfoAsync  # NtUserGetWindowTrackInfoAsync
-2582 stub -noname ScheduleDispatchNotification  # NtUserScheduleDispatchNotification
+2582 stdcall -noname ScheduleDispatchNotification(ptr) NtUserScheduleDispatchNotification
 
 2584 stub -noname EnableModernAppWindowKeyboardIntercept  # NtUserEnableModernAppWindowKeyboardIntercept
 2585 stub -noname UpdateWindowTrackingInfo  # NtUserUpdateWindowTrackingInfo
diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 1fbf8c4a041..1f824f85c50 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2782,3 +2782,16 @@ BOOL WINAPI NtUserRegisterTouchPadCapable( BOOL capable )
     RtlSetLastWin32Error( ERROR_CALL_NOT_IMPLEMENTED );
     return FALSE;
 }
+
+/**********************************************************************
+ *       NtUserScheduleDispatchNotification    (win32u.@)
+ */
+INT WINAPI NtUserScheduleDispatchNotification( HWND hwnd )
+{
+    FIXME("hwnd %p stub!\n", hwnd);
+
+    if (is_window(hwnd))
+        return 2;
+
+    return 0;
+}
diff --git a/dlls/win32u/main.c b/dlls/win32u/main.c
index 3f757e9af21..2fca9c0aa20 100644
--- a/dlls/win32u/main.c
+++ b/dlls/win32u/main.c
@@ -2057,6 +2057,11 @@ BOOL SYSCALL_API NtUserReplyMessage( LRESULT result )
     SYSCALL_FUNC( NtUserReplyMessage );
 }
 
+INT SYSCALL_API NtUserScheduleDispatchNotification( HWND hwnd )
+{
+    SYSCALL_FUNC( NtUserScheduleDispatchNotification );
+}
+
 BOOL SYSCALL_API NtUserScrollDC( HDC hdc, INT dx, INT dy, const RECT *scroll, const RECT *clip,
                                  HRGN ret_update_rgn, RECT *update_rect )
 {
diff --git a/dlls/win32u/win32syscalls.h b/dlls/win32u/win32syscalls.h
index 287e1e42778..fee3a4ef908 100644
--- a/dlls/win32u/win32syscalls.h
+++ b/dlls/win32u/win32syscalls.h
@@ -385,86 +385,87 @@
     SYSCALL_ENTRY( 0x117d, NtUserRemoveMenu, 12 ) \
     SYSCALL_ENTRY( 0x117e, NtUserRemoveProp, 8 ) \
     SYSCALL_ENTRY( 0x117f, NtUserReplyMessage, 4 ) \
-    SYSCALL_ENTRY( 0x1180, NtUserScrollDC, 28 ) \
-    SYSCALL_ENTRY( 0x1181, NtUserScrollWindowEx, 32 ) \
-    SYSCALL_ENTRY( 0x1182, NtUserSelectPalette, 12 ) \
-    SYSCALL_ENTRY( 0x1183, NtUserSendInput, 12 ) \
-    SYSCALL_ENTRY( 0x1184, NtUserSetActiveWindow, 4 ) \
-    SYSCALL_ENTRY( 0x1185, NtUserSetAdditionalForegroundBoostProcesses, 12 ) \
-    SYSCALL_ENTRY( 0x1186, NtUserSetCapture, 4 ) \
-    SYSCALL_ENTRY( 0x1187, NtUserSetCaretBlinkTime, 4 ) \
-    SYSCALL_ENTRY( 0x1188, NtUserSetCaretPos, 8 ) \
-    SYSCALL_ENTRY( 0x1189, NtUserSetClassLong, 16 ) \
-    SYSCALL_ENTRY( 0x118a, NtUserSetClassLongPtr, 16 ) \
-    SYSCALL_ENTRY( 0x118b, NtUserSetClassWord, 12 ) \
-    SYSCALL_ENTRY( 0x118c, NtUserSetClipboardData, 12 ) \
-    SYSCALL_ENTRY( 0x118d, NtUserSetClipboardViewer, 4 ) \
-    SYSCALL_ENTRY( 0x118e, NtUserSetCursor, 4 ) \
-    SYSCALL_ENTRY( 0x118f, NtUserSetCursorIconData, 16 ) \
-    SYSCALL_ENTRY( 0x1190, NtUserSetCursorPos, 8 ) \
-    SYSCALL_ENTRY( 0x1191, NtUserSetFocus, 4 ) \
-    SYSCALL_ENTRY( 0x1192, NtUserSetForegroundWindow, 4 ) \
-    SYSCALL_ENTRY( 0x1193, NtUserSetInternalWindowPos, 16 ) \
-    SYSCALL_ENTRY( 0x1194, NtUserSetKeyboardState, 4 ) \
-    SYSCALL_ENTRY( 0x1195, NtUserSetLayeredWindowAttributes, 16 ) \
-    SYSCALL_ENTRY( 0x1196, NtUserSetMenu, 8 ) \
-    SYSCALL_ENTRY( 0x1197, NtUserSetMenuContextHelpId, 8 ) \
-    SYSCALL_ENTRY( 0x1198, NtUserSetMenuDefaultItem, 12 ) \
-    SYSCALL_ENTRY( 0x1199, NtUserSetObjectInformation, 16 ) \
-    SYSCALL_ENTRY( 0x119a, NtUserSetParent, 8 ) \
-    SYSCALL_ENTRY( 0x119b, NtUserSetProcessDefaultLayout, 4 ) \
-    SYSCALL_ENTRY( 0x119c, NtUserSetProcessDpiAwarenessContext, 8 ) \
-    SYSCALL_ENTRY( 0x119d, NtUserSetProcessWindowStation, 4 ) \
-    SYSCALL_ENTRY( 0x119e, NtUserSetProgmanWindow, 4 ) \
-    SYSCALL_ENTRY( 0x119f, NtUserSetProp, 12 ) \
-    SYSCALL_ENTRY( 0x11a0, NtUserSetScrollInfo, 16 ) \
-    SYSCALL_ENTRY( 0x11a1, NtUserSetShellWindowEx, 8 ) \
-    SYSCALL_ENTRY( 0x11a2, NtUserSetSysColors, 12 ) \
-    SYSCALL_ENTRY( 0x11a3, NtUserSetSystemMenu, 8 ) \
-    SYSCALL_ENTRY( 0x11a4, NtUserSetSystemTimer, 12 ) \
-    SYSCALL_ENTRY( 0x11a5, NtUserSetTaskmanWindow, 4 ) \
-    SYSCALL_ENTRY( 0x11a6, NtUserSetThreadDesktop, 4 ) \
-    SYSCALL_ENTRY( 0x11a7, NtUserSetTimer, 20 ) \
-    SYSCALL_ENTRY( 0x11a8, NtUserSetWinEventHook, 32 ) \
-    SYSCALL_ENTRY( 0x11a9, NtUserSetWindowContextHelpId, 8 ) \
-    SYSCALL_ENTRY( 0x11aa, NtUserSetWindowLong, 16 ) \
-    SYSCALL_ENTRY( 0x11ab, NtUserSetWindowLongPtr, 16 ) \
-    SYSCALL_ENTRY( 0x11ac, NtUserSetWindowPlacement, 8 ) \
-    SYSCALL_ENTRY( 0x11ad, NtUserSetWindowPos, 28 ) \
-    SYSCALL_ENTRY( 0x11ae, NtUserSetWindowRgn, 12 ) \
-    SYSCALL_ENTRY( 0x11af, NtUserSetWindowWord, 12 ) \
-    SYSCALL_ENTRY( 0x11b0, NtUserSetWindowsHookEx, 24 ) \
-    SYSCALL_ENTRY( 0x11b1, NtUserShowCaret, 4 ) \
-    SYSCALL_ENTRY( 0x11b2, NtUserShowCursor, 4 ) \
-    SYSCALL_ENTRY( 0x11b3, NtUserShowOwnedPopups, 8 ) \
-    SYSCALL_ENTRY( 0x11b4, NtUserShowScrollBar, 12 ) \
-    SYSCALL_ENTRY( 0x11b5, NtUserShowWindow, 8 ) \
-    SYSCALL_ENTRY( 0x11b6, NtUserShowWindowAsync, 8 ) \
-    SYSCALL_ENTRY( 0x11b7, NtUserSwitchDesktop, 4 ) \
-    SYSCALL_ENTRY( 0x11b8, NtUserSystemParametersInfo, 16 ) \
-    SYSCALL_ENTRY( 0x11b9, NtUserSystemParametersInfoForDpi, 20 ) \
-    SYSCALL_ENTRY( 0x11ba, NtUserThunkedMenuInfo, 8 ) \
-    SYSCALL_ENTRY( 0x11bb, NtUserThunkedMenuItemInfo, 24 ) \
-    SYSCALL_ENTRY( 0x11bc, NtUserToUnicodeEx, 28 ) \
-    SYSCALL_ENTRY( 0x11bd, NtUserTrackMouseEvent, 4 ) \
-    SYSCALL_ENTRY( 0x11be, NtUserTrackPopupMenuEx, 24 ) \
-    SYSCALL_ENTRY( 0x11bf, NtUserTranslateAccelerator, 12 ) \
-    SYSCALL_ENTRY( 0x11c0, NtUserTranslateMessage, 8 ) \
-    SYSCALL_ENTRY( 0x11c1, NtUserUnhookWinEvent, 4 ) \
-    SYSCALL_ENTRY( 0x11c2, NtUserUnhookWindowsHook, 8 ) \
-    SYSCALL_ENTRY( 0x11c3, NtUserUnhookWindowsHookEx, 4 ) \
-    SYSCALL_ENTRY( 0x11c4, NtUserUnregisterClass, 12 ) \
-    SYSCALL_ENTRY( 0x11c5, NtUserUnregisterHotKey, 8 ) \
-    SYSCALL_ENTRY( 0x11c6, NtUserUpdateInputContext, 12 ) \
-    SYSCALL_ENTRY( 0x11c7, NtUserUpdateLayeredWindow, 40 ) \
-    SYSCALL_ENTRY( 0x11c8, NtUserValidateRect, 8 ) \
-    SYSCALL_ENTRY( 0x11c9, NtUserValidateRgn, 8 ) \
-    SYSCALL_ENTRY( 0x11ca, NtUserVkKeyScanEx, 8 ) \
-    SYSCALL_ENTRY( 0x11cb, NtUserWaitForInputIdle, 12 ) \
-    SYSCALL_ENTRY( 0x11cc, NtUserWaitMessage, 0 ) \
-    SYSCALL_ENTRY( 0x11cd, NtUserWindowFromDC, 4 ) \
-    SYSCALL_ENTRY( 0x11ce, NtUserWindowFromPoint, 8 ) \
-    SYSCALL_ENTRY( 0x11cf, __wine_get_icm_profile, 16 )
+    SYSCALL_ENTRY( 0x1180, NtUserScheduleDispatchNotification, 4 ) \
+    SYSCALL_ENTRY( 0x1181, NtUserScrollDC, 28 ) \
+    SYSCALL_ENTRY( 0x1182, NtUserScrollWindowEx, 32 ) \
+    SYSCALL_ENTRY( 0x1183, NtUserSelectPalette, 12 ) \
+    SYSCALL_ENTRY( 0x1184, NtUserSendInput, 12 ) \
+    SYSCALL_ENTRY( 0x1185, NtUserSetActiveWindow, 4 ) \
+    SYSCALL_ENTRY( 0x1186, NtUserSetAdditionalForegroundBoostProcesses, 12 ) \
+    SYSCALL_ENTRY( 0x1187, NtUserSetCapture, 4 ) \
+    SYSCALL_ENTRY( 0x1188, NtUserSetCaretBlinkTime, 4 ) \
+    SYSCALL_ENTRY( 0x1189, NtUserSetCaretPos, 8 ) \
+    SYSCALL_ENTRY( 0x118a, NtUserSetClassLong, 16 ) \
+    SYSCALL_ENTRY( 0x118b, NtUserSetClassLongPtr, 16 ) \
+    SYSCALL_ENTRY( 0x118c, NtUserSetClassWord, 12 ) \
+    SYSCALL_ENTRY( 0x118d, NtUserSetClipboardData, 12 ) \
+    SYSCALL_ENTRY( 0x118e, NtUserSetClipboardViewer, 4 ) \
+    SYSCALL_ENTRY( 0x118f, NtUserSetCursor, 4 ) \
+    SYSCALL_ENTRY( 0x1190, NtUserSetCursorIconData, 16 ) \
+    SYSCALL_ENTRY( 0x1191, NtUserSetCursorPos, 8 ) \
+    SYSCALL_ENTRY( 0x1192, NtUserSetFocus, 4 ) \
+    SYSCALL_ENTRY( 0x1193, NtUserSetForegroundWindow, 4 ) \
+    SYSCALL_ENTRY( 0x1194, NtUserSetInternalWindowPos, 16 ) \
+    SYSCALL_ENTRY( 0x1195, NtUserSetKeyboardState, 4 ) \
+    SYSCALL_ENTRY( 0x1196, NtUserSetLayeredWindowAttributes, 16 ) \
+    SYSCALL_ENTRY( 0x1197, NtUserSetMenu, 8 ) \
+    SYSCALL_ENTRY( 0x1198, NtUserSetMenuContextHelpId, 8 ) \
+    SYSCALL_ENTRY( 0x1199, NtUserSetMenuDefaultItem, 12 ) \
+    SYSCALL_ENTRY( 0x119a, NtUserSetObjectInformation, 16 ) \
+    SYSCALL_ENTRY( 0x119b, NtUserSetParent, 8 ) \
+    SYSCALL_ENTRY( 0x119c, NtUserSetProcessDefaultLayout, 4 ) \
+    SYSCALL_ENTRY( 0x119d, NtUserSetProcessDpiAwarenessContext, 8 ) \
+    SYSCALL_ENTRY( 0x119e, NtUserSetProcessWindowStation, 4 ) \
+    SYSCALL_ENTRY( 0x119f, NtUserSetProgmanWindow, 4 ) \
+    SYSCALL_ENTRY( 0x11a0, NtUserSetProp, 12 ) \
+    SYSCALL_ENTRY( 0x11a1, NtUserSetScrollInfo, 16 ) \
+    SYSCALL_ENTRY( 0x11a2, NtUserSetShellWindowEx, 8 ) \
+    SYSCALL_ENTRY( 0x11a3, NtUserSetSysColors, 12 ) \
+    SYSCALL_ENTRY( 0x11a4, NtUserSetSystemMenu, 8 ) \
+    SYSCALL_ENTRY( 0x11a5, NtUserSetSystemTimer, 12 ) \
+    SYSCALL_ENTRY( 0x11a6, NtUserSetTaskmanWindow, 4 ) \
+    SYSCALL_ENTRY( 0x11a7, NtUserSetThreadDesktop, 4 ) \
+    SYSCALL_ENTRY( 0x11a8, NtUserSetTimer, 20 ) \
+    SYSCALL_ENTRY( 0x11a9, NtUserSetWinEventHook, 32 ) \
+    SYSCALL_ENTRY( 0x11aa, NtUserSetWindowContextHelpId, 8 ) \
+    SYSCALL_ENTRY( 0x11ab, NtUserSetWindowLong, 16 ) \
+    SYSCALL_ENTRY( 0x11ac, NtUserSetWindowLongPtr, 16 ) \
+    SYSCALL_ENTRY( 0x11ad, NtUserSetWindowPlacement, 8 ) \
+    SYSCALL_ENTRY( 0x11ae, NtUserSetWindowPos, 28 ) \
+    SYSCALL_ENTRY( 0x11af, NtUserSetWindowRgn, 12 ) \
+    SYSCALL_ENTRY( 0x11b0, NtUserSetWindowWord, 12 ) \
+    SYSCALL_ENTRY( 0x11b1, NtUserSetWindowsHookEx, 24 ) \
+    SYSCALL_ENTRY( 0x11b2, NtUserShowCaret, 4 ) \
+    SYSCALL_ENTRY( 0x11b3, NtUserShowCursor, 4 ) \
+    SYSCALL_ENTRY( 0x11b4, NtUserShowOwnedPopups, 8 ) \
+    SYSCALL_ENTRY( 0x11b5, NtUserShowScrollBar, 12 ) \
+    SYSCALL_ENTRY( 0x11b6, NtUserShowWindow, 8 ) \
+    SYSCALL_ENTRY( 0x11b7, NtUserShowWindowAsync, 8 ) \
+    SYSCALL_ENTRY( 0x11b8, NtUserSwitchDesktop, 4 ) \
+    SYSCALL_ENTRY( 0x11b9, NtUserSystemParametersInfo, 16 ) \
+    SYSCALL_ENTRY( 0x11ba, NtUserSystemParametersInfoForDpi, 20 ) \
+    SYSCALL_ENTRY( 0x11bb, NtUserThunkedMenuInfo, 8 ) \
+    SYSCALL_ENTRY( 0x11bc, NtUserThunkedMenuItemInfo, 24 ) \
+    SYSCALL_ENTRY( 0x11bd, NtUserToUnicodeEx, 28 ) \
+    SYSCALL_ENTRY( 0x11be, NtUserTrackMouseEvent, 4 ) \
+    SYSCALL_ENTRY( 0x11bf, NtUserTrackPopupMenuEx, 24 ) \
+    SYSCALL_ENTRY( 0x11c0, NtUserTranslateAccelerator, 12 ) \
+    SYSCALL_ENTRY( 0x11c1, NtUserTranslateMessage, 8 ) \
+    SYSCALL_ENTRY( 0x11c2, NtUserUnhookWinEvent, 4 ) \
+    SYSCALL_ENTRY( 0x11c3, NtUserUnhookWindowsHook, 8 ) \
+    SYSCALL_ENTRY( 0x11c4, NtUserUnhookWindowsHookEx, 4 ) \
+    SYSCALL_ENTRY( 0x11c5, NtUserUnregisterClass, 12 ) \
+    SYSCALL_ENTRY( 0x11c6, NtUserUnregisterHotKey, 8 ) \
+    SYSCALL_ENTRY( 0x11c7, NtUserUpdateInputContext, 12 ) \
+    SYSCALL_ENTRY( 0x11c8, NtUserUpdateLayeredWindow, 40 ) \
+    SYSCALL_ENTRY( 0x11c9, NtUserValidateRect, 8 ) \
+    SYSCALL_ENTRY( 0x11ca, NtUserValidateRgn, 8 ) \
+    SYSCALL_ENTRY( 0x11cb, NtUserVkKeyScanEx, 8 ) \
+    SYSCALL_ENTRY( 0x11cc, NtUserWaitForInputIdle, 12 ) \
+    SYSCALL_ENTRY( 0x11cd, NtUserWaitMessage, 0 ) \
+    SYSCALL_ENTRY( 0x11ce, NtUserWindowFromDC, 4 ) \
+    SYSCALL_ENTRY( 0x11cf, NtUserWindowFromPoint, 8 ) \
+    SYSCALL_ENTRY( 0x11d0, __wine_get_icm_profile, 16 )
 #ifdef _WIN64
 #define ALL_SYSCALLS \
     SYSCALL_ENTRY( 0x1000, NtGdiAbortDoc, 8 ) \
@@ -851,86 +852,87 @@
     SYSCALL_ENTRY( 0x117d, NtUserRemoveMenu, 24 ) \
     SYSCALL_ENTRY( 0x117e, NtUserRemoveProp, 16 ) \
     SYSCALL_ENTRY( 0x117f, NtUserReplyMessage, 8 ) \
-    SYSCALL_ENTRY( 0x1180, NtUserScrollDC, 56 ) \
-    SYSCALL_ENTRY( 0x1181, NtUserScrollWindowEx, 64 ) \
-    SYSCALL_ENTRY( 0x1182, NtUserSelectPalette, 24 ) \
-    SYSCALL_ENTRY( 0x1183, NtUserSendInput, 24 ) \
-    SYSCALL_ENTRY( 0x1184, NtUserSetActiveWindow, 8 ) \
-    SYSCALL_ENTRY( 0x1185, NtUserSetAdditionalForegroundBoostProcesses, 24 ) \
-    SYSCALL_ENTRY( 0x1186, NtUserSetCapture, 8 ) \
-    SYSCALL_ENTRY( 0x1187, NtUserSetCaretBlinkTime, 8 ) \
-    SYSCALL_ENTRY( 0x1188, NtUserSetCaretPos, 16 ) \
-    SYSCALL_ENTRY( 0x1189, NtUserSetClassLong, 32 ) \
-    SYSCALL_ENTRY( 0x118a, NtUserSetClassLongPtr, 32 ) \
-    SYSCALL_ENTRY( 0x118b, NtUserSetClassWord, 24 ) \
-    SYSCALL_ENTRY( 0x118c, NtUserSetClipboardData, 24 ) \
-    SYSCALL_ENTRY( 0x118d, NtUserSetClipboardViewer, 8 ) \
-    SYSCALL_ENTRY( 0x118e, NtUserSetCursor, 8 ) \
-    SYSCALL_ENTRY( 0x118f, NtUserSetCursorIconData, 32 ) \
-    SYSCALL_ENTRY( 0x1190, NtUserSetCursorPos, 16 ) \
-    SYSCALL_ENTRY( 0x1191, NtUserSetFocus, 8 ) \
-    SYSCALL_ENTRY( 0x1192, NtUserSetForegroundWindow, 8 ) \
-    SYSCALL_ENTRY( 0x1193, NtUserSetInternalWindowPos, 32 ) \
-    SYSCALL_ENTRY( 0x1194, NtUserSetKeyboardState, 8 ) \
-    SYSCALL_ENTRY( 0x1195, NtUserSetLayeredWindowAttributes, 32 ) \
-    SYSCALL_ENTRY( 0x1196, NtUserSetMenu, 16 ) \
-    SYSCALL_ENTRY( 0x1197, NtUserSetMenuContextHelpId, 16 ) \
-    SYSCALL_ENTRY( 0x1198, NtUserSetMenuDefaultItem, 24 ) \
-    SYSCALL_ENTRY( 0x1199, NtUserSetObjectInformation, 32 ) \
-    SYSCALL_ENTRY( 0x119a, NtUserSetParent, 16 ) \
-    SYSCALL_ENTRY( 0x119b, NtUserSetProcessDefaultLayout, 8 ) \
-    SYSCALL_ENTRY( 0x119c, NtUserSetProcessDpiAwarenessContext, 16 ) \
-    SYSCALL_ENTRY( 0x119d, NtUserSetProcessWindowStation, 8 ) \
-    SYSCALL_ENTRY( 0x119e, NtUserSetProgmanWindow, 8 ) \
-    SYSCALL_ENTRY( 0x119f, NtUserSetProp, 24 ) \
-    SYSCALL_ENTRY( 0x11a0, NtUserSetScrollInfo, 32 ) \
-    SYSCALL_ENTRY( 0x11a1, NtUserSetShellWindowEx, 16 ) \
-    SYSCALL_ENTRY( 0x11a2, NtUserSetSysColors, 24 ) \
-    SYSCALL_ENTRY( 0x11a3, NtUserSetSystemMenu, 16 ) \
-    SYSCALL_ENTRY( 0x11a4, NtUserSetSystemTimer, 24 ) \
-    SYSCALL_ENTRY( 0x11a5, NtUserSetTaskmanWindow, 8 ) \
-    SYSCALL_ENTRY( 0x11a6, NtUserSetThreadDesktop, 8 ) \
-    SYSCALL_ENTRY( 0x11a7, NtUserSetTimer, 40 ) \
-    SYSCALL_ENTRY( 0x11a8, NtUserSetWinEventHook, 64 ) \
-    SYSCALL_ENTRY( 0x11a9, NtUserSetWindowContextHelpId, 16 ) \
-    SYSCALL_ENTRY( 0x11aa, NtUserSetWindowLong, 32 ) \
-    SYSCALL_ENTRY( 0x11ab, NtUserSetWindowLongPtr, 32 ) \
-    SYSCALL_ENTRY( 0x11ac, NtUserSetWindowPlacement, 16 ) \
-    SYSCALL_ENTRY( 0x11ad, NtUserSetWindowPos, 56 ) \
-    SYSCALL_ENTRY( 0x11ae, NtUserSetWindowRgn, 24 ) \
-    SYSCALL_ENTRY( 0x11af, NtUserSetWindowWord, 24 ) \
-    SYSCALL_ENTRY( 0x11b0, NtUserSetWindowsHookEx, 48 ) \
-    SYSCALL_ENTRY( 0x11b1, NtUserShowCaret, 8 ) \
-    SYSCALL_ENTRY( 0x11b2, NtUserShowCursor, 8 ) \
-    SYSCALL_ENTRY( 0x11b3, NtUserShowOwnedPopups, 16 ) \
-    SYSCALL_ENTRY( 0x11b4, NtUserShowScrollBar, 24 ) \
-    SYSCALL_ENTRY( 0x11b5, NtUserShowWindow, 16 ) \
-    SYSCALL_ENTRY( 0x11b6, NtUserShowWindowAsync, 16 ) \
-    SYSCALL_ENTRY( 0x11b7, NtUserSwitchDesktop, 8 ) \
-    SYSCALL_ENTRY( 0x11b8, NtUserSystemParametersInfo, 32 ) \
-    SYSCALL_ENTRY( 0x11b9, NtUserSystemParametersInfoForDpi, 40 ) \
-    SYSCALL_ENTRY( 0x11ba, NtUserThunkedMenuInfo, 16 ) \
-    SYSCALL_ENTRY( 0x11bb, NtUserThunkedMenuItemInfo, 48 ) \
-    SYSCALL_ENTRY( 0x11bc, NtUserToUnicodeEx, 56 ) \
-    SYSCALL_ENTRY( 0x11bd, NtUserTrackMouseEvent, 8 ) \
-    SYSCALL_ENTRY( 0x11be, NtUserTrackPopupMenuEx, 48 ) \
-    SYSCALL_ENTRY( 0x11bf, NtUserTranslateAccelerator, 24 ) \
-    SYSCALL_ENTRY( 0x11c0, NtUserTranslateMessage, 16 ) \
-    SYSCALL_ENTRY( 0x11c1, NtUserUnhookWinEvent, 8 ) \
-    SYSCALL_ENTRY( 0x11c2, NtUserUnhookWindowsHook, 16 ) \
-    SYSCALL_ENTRY( 0x11c3, NtUserUnhookWindowsHookEx, 8 ) \
-    SYSCALL_ENTRY( 0x11c4, NtUserUnregisterClass, 24 ) \
-    SYSCALL_ENTRY( 0x11c5, NtUserUnregisterHotKey, 16 ) \
-    SYSCALL_ENTRY( 0x11c6, NtUserUpdateInputContext, 24 ) \
-    SYSCALL_ENTRY( 0x11c7, NtUserUpdateLayeredWindow, 80 ) \
-    SYSCALL_ENTRY( 0x11c8, NtUserValidateRect, 16 ) \
-    SYSCALL_ENTRY( 0x11c9, NtUserValidateRgn, 16 ) \
-    SYSCALL_ENTRY( 0x11ca, NtUserVkKeyScanEx, 16 ) \
-    SYSCALL_ENTRY( 0x11cb, NtUserWaitForInputIdle, 24 ) \
-    SYSCALL_ENTRY( 0x11cc, NtUserWaitMessage, 0 ) \
-    SYSCALL_ENTRY( 0x11cd, NtUserWindowFromDC, 8 ) \
-    SYSCALL_ENTRY( 0x11ce, NtUserWindowFromPoint, 16 ) \
-    SYSCALL_ENTRY( 0x11cf, __wine_get_icm_profile, 32 )
+    SYSCALL_ENTRY( 0x1180, NtUserScheduleDispatchNotification, 8 ) \
+    SYSCALL_ENTRY( 0x1181, NtUserScrollDC, 56 ) \
+    SYSCALL_ENTRY( 0x1182, NtUserScrollWindowEx, 64 ) \
+    SYSCALL_ENTRY( 0x1183, NtUserSelectPalette, 24 ) \
+    SYSCALL_ENTRY( 0x1184, NtUserSendInput, 24 ) \
+    SYSCALL_ENTRY( 0x1185, NtUserSetActiveWindow, 8 ) \
+    SYSCALL_ENTRY( 0x1186, NtUserSetAdditionalForegroundBoostProcesses, 24 ) \
+    SYSCALL_ENTRY( 0x1187, NtUserSetCapture, 8 ) \
+    SYSCALL_ENTRY( 0x1188, NtUserSetCaretBlinkTime, 8 ) \
+    SYSCALL_ENTRY( 0x1189, NtUserSetCaretPos, 16 ) \
+    SYSCALL_ENTRY( 0x118a, NtUserSetClassLong, 32 ) \
+    SYSCALL_ENTRY( 0x118b, NtUserSetClassLongPtr, 32 ) \
+    SYSCALL_ENTRY( 0x118c, NtUserSetClassWord, 24 ) \
+    SYSCALL_ENTRY( 0x118d, NtUserSetClipboardData, 24 ) \
+    SYSCALL_ENTRY( 0x118e, NtUserSetClipboardViewer, 8 ) \
+    SYSCALL_ENTRY( 0x118f, NtUserSetCursor, 8 ) \
+    SYSCALL_ENTRY( 0x1190, NtUserSetCursorIconData, 32 ) \
+    SYSCALL_ENTRY( 0x1191, NtUserSetCursorPos, 16 ) \
+    SYSCALL_ENTRY( 0x1192, NtUserSetFocus, 8 ) \
+    SYSCALL_ENTRY( 0x1193, NtUserSetForegroundWindow, 8 ) \
+    SYSCALL_ENTRY( 0x1194, NtUserSetInternalWindowPos, 32 ) \
+    SYSCALL_ENTRY( 0x1195, NtUserSetKeyboardState, 8 ) \
+    SYSCALL_ENTRY( 0x1196, NtUserSetLayeredWindowAttributes, 32 ) \
+    SYSCALL_ENTRY( 0x1197, NtUserSetMenu, 16 ) \
+    SYSCALL_ENTRY( 0x1198, NtUserSetMenuContextHelpId, 16 ) \
+    SYSCALL_ENTRY( 0x1199, NtUserSetMenuDefaultItem, 24 ) \
+    SYSCALL_ENTRY( 0x119a, NtUserSetObjectInformation, 32 ) \
+    SYSCALL_ENTRY( 0x119b, NtUserSetParent, 16 ) \
+    SYSCALL_ENTRY( 0x119c, NtUserSetProcessDefaultLayout, 8 ) \
+    SYSCALL_ENTRY( 0x119d, NtUserSetProcessDpiAwarenessContext, 16 ) \
+    SYSCALL_ENTRY( 0x119e, NtUserSetProcessWindowStation, 8 ) \
+    SYSCALL_ENTRY( 0x119f, NtUserSetProgmanWindow, 8 ) \
+    SYSCALL_ENTRY( 0x11a0, NtUserSetProp, 24 ) \
+    SYSCALL_ENTRY( 0x11a1, NtUserSetScrollInfo, 32 ) \
+    SYSCALL_ENTRY( 0x11a2, NtUserSetShellWindowEx, 16 ) \
+    SYSCALL_ENTRY( 0x11a3, NtUserSetSysColors, 24 ) \
+    SYSCALL_ENTRY( 0x11a4, NtUserSetSystemMenu, 16 ) \
+    SYSCALL_ENTRY( 0x11a5, NtUserSetSystemTimer, 24 ) \
+    SYSCALL_ENTRY( 0x11a6, NtUserSetTaskmanWindow, 8 ) \
+    SYSCALL_ENTRY( 0x11a7, NtUserSetThreadDesktop, 8 ) \
+    SYSCALL_ENTRY( 0x11a8, NtUserSetTimer, 40 ) \
+    SYSCALL_ENTRY( 0x11a9, NtUserSetWinEventHook, 64 ) \
+    SYSCALL_ENTRY( 0x11aa, NtUserSetWindowContextHelpId, 16 ) \
+    SYSCALL_ENTRY( 0x11ab, NtUserSetWindowLong, 32 ) \
+    SYSCALL_ENTRY( 0x11ac, NtUserSetWindowLongPtr, 32 ) \
+    SYSCALL_ENTRY( 0x11ad, NtUserSetWindowPlacement, 16 ) \
+    SYSCALL_ENTRY( 0x11ae, NtUserSetWindowPos, 56 ) \
+    SYSCALL_ENTRY( 0x11af, NtUserSetWindowRgn, 24 ) \
+    SYSCALL_ENTRY( 0x11b0, NtUserSetWindowWord, 24 ) \
+    SYSCALL_ENTRY( 0x11b1, NtUserSetWindowsHookEx, 48 ) \
+    SYSCALL_ENTRY( 0x11b2, NtUserShowCaret, 8 ) \
+    SYSCALL_ENTRY( 0x11b3, NtUserShowCursor, 8 ) \
+    SYSCALL_ENTRY( 0x11b4, NtUserShowOwnedPopups, 16 ) \
+    SYSCALL_ENTRY( 0x11b5, NtUserShowScrollBar, 24 ) \
+    SYSCALL_ENTRY( 0x11b6, NtUserShowWindow, 16 ) \
+    SYSCALL_ENTRY( 0x11b7, NtUserShowWindowAsync, 16 ) \
+    SYSCALL_ENTRY( 0x11b8, NtUserSwitchDesktop, 8 ) \
+    SYSCALL_ENTRY( 0x11b9, NtUserSystemParametersInfo, 32 ) \
+    SYSCALL_ENTRY( 0x11ba, NtUserSystemParametersInfoForDpi, 40 ) \
+    SYSCALL_ENTRY( 0x11bb, NtUserThunkedMenuInfo, 16 ) \
+    SYSCALL_ENTRY( 0x11bc, NtUserThunkedMenuItemInfo, 48 ) \
+    SYSCALL_ENTRY( 0x11bd, NtUserToUnicodeEx, 56 ) \
+    SYSCALL_ENTRY( 0x11be, NtUserTrackMouseEvent, 8 ) \
+    SYSCALL_ENTRY( 0x11bf, NtUserTrackPopupMenuEx, 48 ) \
+    SYSCALL_ENTRY( 0x11c0, NtUserTranslateAccelerator, 24 ) \
+    SYSCALL_ENTRY( 0x11c1, NtUserTranslateMessage, 16 ) \
+    SYSCALL_ENTRY( 0x11c2, NtUserUnhookWinEvent, 8 ) \
+    SYSCALL_ENTRY( 0x11c3, NtUserUnhookWindowsHook, 16 ) \
+    SYSCALL_ENTRY( 0x11c4, NtUserUnhookWindowsHookEx, 8 ) \
+    SYSCALL_ENTRY( 0x11c5, NtUserUnregisterClass, 24 ) \
+    SYSCALL_ENTRY( 0x11c6, NtUserUnregisterHotKey, 16 ) \
+    SYSCALL_ENTRY( 0x11c7, NtUserUpdateInputContext, 24 ) \
+    SYSCALL_ENTRY( 0x11c8, NtUserUpdateLayeredWindow, 80 ) \
+    SYSCALL_ENTRY( 0x11c9, NtUserValidateRect, 16 ) \
+    SYSCALL_ENTRY( 0x11ca, NtUserValidateRgn, 16 ) \
+    SYSCALL_ENTRY( 0x11cb, NtUserVkKeyScanEx, 16 ) \
+    SYSCALL_ENTRY( 0x11cc, NtUserWaitForInputIdle, 24 ) \
+    SYSCALL_ENTRY( 0x11cd, NtUserWaitMessage, 0 ) \
+    SYSCALL_ENTRY( 0x11ce, NtUserWindowFromDC, 8 ) \
+    SYSCALL_ENTRY( 0x11cf, NtUserWindowFromPoint, 16 ) \
+    SYSCALL_ENTRY( 0x11d0, __wine_get_icm_profile, 32 )
 #else
 #define ALL_SYSCALLS ALL_SYSCALLS32
 #endif
diff --git a/dlls/win32u/win32u.spec b/dlls/win32u/win32u.spec
index 9e25f3d8bba..09101ffd779 100644
--- a/dlls/win32u/win32u.spec
+++ b/dlls/win32u/win32u.spec
@@ -1320,7 +1320,7 @@
 @ stub NtUserRestoreWindowDpiChanges
 @ stub NtUserSBGetParms
 @ stub NtUserScaleSystemMetricForDPIWithoutCache
-@ stub NtUserScheduleDispatchNotification
+@ stdcall -syscall NtUserScheduleDispatchNotification(ptr)
 @ stdcall -syscall NtUserScrollDC(long long long ptr ptr long ptr)
 @ stdcall -syscall NtUserScrollWindowEx(long long long ptr ptr long ptr long)
 @ stdcall -syscall NtUserSelectPalette(long long long)
diff --git a/dlls/wow64win/user.c b/dlls/wow64win/user.c
index 594085e3777..ee51488dfd5 100644
--- a/dlls/wow64win/user.c
+++ b/dlls/wow64win/user.c
@@ -4086,6 +4086,13 @@ NTSTATUS WINAPI wow64_NtUserReplyMessage( UINT *args )
     return NtUserReplyMessage( result );
 }
 
+NTSTATUS WINAPI wow64_NtUserScheduleDispatchNotification( UINT *args )
+{
+    HWND hwnd = get_handle( &args );
+
+    return NtUserScheduleDispatchNotification( hwnd );
+}
+
 NTSTATUS WINAPI wow64_NtUserScrollDC( UINT *args )
 {
     HDC hdc = get_handle( &args );
diff --git a/include/ntuser.h b/include/ntuser.h
index 8e539bb1d69..de63e93a352 100644
--- a/include/ntuser.h
+++ b/include/ntuser.h
@@ -956,6 +956,7 @@ W32KAPI BOOL    WINAPI NtUserRemoveClipboardFormatListener( HWND hwnd );
 W32KAPI BOOL    WINAPI NtUserRemoveMenu( HMENU menu, UINT id, UINT flags );
 W32KAPI HANDLE  WINAPI NtUserRemoveProp( HWND hwnd, const WCHAR *str );
 W32KAPI BOOL    WINAPI NtUserReplyMessage( LRESULT result );
+W32KAPI INT     WINAPI NtUserScheduleDispatchNotification( HWND hwnd );
 W32KAPI BOOL    WINAPI NtUserScrollDC( HDC hdc, INT dx, INT dy, const RECT *scroll, const RECT *clip,
                                        HRGN ret_update_rgn, RECT *update_rect );
 W32KAPI INT     WINAPI NtUserScrollWindowEx( HWND hwnd, INT dx, INT dy, const RECT *rect,
-- 
2.43.0

